atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -51,80 +55,102 @@
     private static final String VALUE_COLUMN = "value";
 
     private static final byte[] SUCCESSFUL_OPERATION = {1};
 
-    private static final String BACKUP_COLUMN_NAME = "oldTs";
-    private static final byte[] INVALIDATED_VALUE = new byte[1];
+    public static final String BACKUP_COLUMN_NAME = "oldTs";
+
+    static final byte[] INVALIDATED_VALUE = new byte[1];
 
     private CassandraTimestampUtils() {
         // utility class
     }
 
-    public static ByteBuffer constructSelectTimestampBoundQuery() {
-        String selectQuery = String.format(
-                "SELECT %s FROM %s WHERE key=%s AND column1=%s AND column2=%s;",
+    public static ByteBuffer constructSelectFromTimestampTableQuery() {
+        return toByteBuffer(String.format(
+                "SELECT %s, %s FROM %s WHERE key=%s;",
+                COLUMN_NAME_COLUMN,
                 VALUE_COLUMN,
                 wrapInQuotes(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName()),
-                ROW_AND_COLUMN_NAME_HEX_STRING,
-                ROW_AND_COLUMN_NAME_HEX_STRING,
-                CASSANDRA_TIMESTAMP);
-        return ByteBuffer.wrap(PtBytes.toBytes(selectQuery));
+                ROW_AND_COLUMN_NAME_HEX_STRING));
     }
 
-    public static ByteBuffer constructCheckAndSetCqlQuery(Long expected, long target) {
-        String updateQuery = (expected == null)
-                ? constructInsertIfNotExistsQuery(target)
-                : constructUpdateIfEqualQuery(expected, target);
+    public static ByteBuffer constructCheckAndSetMultipleQuery(Map<String, Pair<byte[], byte[]>> checkAndSetRequest) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("BEGIN UNLOGGED BATCH\n"); // Safe, because all updates are on the same partition key
+
+        // Safe, because ordering does not apply in batches
+        checkAndSetRequest.entrySet().forEach(entry -> {
+            String columnName = entry.getKey();
+            byte[] expected = entry.getValue().getLhSide();
+            byte[] target = entry.getValue().getRhSide();
+            builder.append(constructCheckAndSetQuery(columnName, expected, target));
+        });
+        builder.append("APPLY BATCH;");
+        return toByteBuffer(builder.toString());
+    }
 
-        return ByteBuffer.wrap(PtBytes.toBytes(updateQuery));
+    private static String constructCheckAndSetQuery(String columnName, byte[] expected, byte[] target) {
+        if (expected == null) {
+            Preconditions.checkArgument(target != null, "[CAS] At least one of expected and target must be provided!");
+            return constructInsertIfNotExistsQuery(columnName, target);
+        }
+        if (target == null) {
+            // We treat this as a no op for now, which satisfies our purposes.
+            return "";
+        }
+        return constructUpdateIfEqualQuery(columnName, expected, target);
     }
 
     public static boolean wasOperationApplied(CqlResult cqlResult) {
```

Comment:
Ah, I forgot to cleanup after an extract-method. I've reordered the methods accordingly. (For order of appearance there doesn't seem to be an easy way to do it - I did a pass through on find-usages which hopefully should be pretty close, though the ordering isn't necessarily stable if you have private methods only called by other private methods!)